sdks/java/io/cassandra/src/main/java/org/apache/beam/sdk/io/cassandra/CassandraServiceImpl.java

```
@@ -0,0 +1,318 @@
+      resultSet = session.execute(splitQuery);
+
+      final MappingManager mappingManager = new MappingManager(session);
+      Mapper mapper = mappingManager.mapper(entity);
+      iterator = mapper.map(resultSet).iterator();
+      return advance();
+    }
+
+    @Override
+    public boolean advance() throws IOException {
+      if (iterator.hasNext()) {
+        current = iterator.next();
+        return true;
+      }
+      current = null;
+      return false;
+    }
+
+    @Override
+    public void close() {
+      LOG.debug("Closing Cassandra reader");
+      if (session != null) {
+        session.close();
+      }
+      if (cluster != null) {
+        cluster.close();
+      }
+    }
+
+    @Override
+    public T getCurrent() throws NoSuchElementException {
+      if (current == null) {
+        throw new NoSuchElementException();
+      }
+      return current;
+    }
+  }
+
+  @Override
+  public Reader<T> createReader(String query, Class<T> entity) throws IOException {
+    return new CassandraReaderImpl<>(query, entity);
+  }
+
+  @Override
+  public long getEstimatedSizeBytes() {
+    try (Cluster cluster = connectionConfiguration.getCluster()) {
+      List<TokenRange> tokenRanges = getTokenRanges(cluster, connectionConfiguration, table);
+      long size = 0L;
+      for (TokenRange tokenRange : tokenRanges) {
+        size += tokenRange.meanPartitionSize * tokenRange.partitionCount;
+      }
+      return Math.round(size / getRingFraction(cluster, tokenRanges));
+    }
+  }
+
+  @Override
+  public List<BoundedSource<T>> splitIntoBundles(
+      CassandraIO.Read<T> spec,
+      long desiredBundleSizeBytes,
+      SerializableFunction<PipelineOptions, CassandraService<T>> serviceFactory) {
+    long numSplits = 1;
+    List<BoundedSource<T>> sourceList = new ArrayList<>();
+    try {
+      if (desiredBundleSizeBytes > 0) {
+        numSplits = getEstimatedSizeBytes() / desiredBundleSizeBytes;
+      }
+    } catch (Exception e) {
```

Do not silently swallow exceptions.

---

I added a log message (`warn`) in case of exception.